### PR TITLE
[universal] Update npm due to GHSA-c2qf-rxjj-qqgw

### PR DESCRIPTION
**Pull request by alexander-smolyakov**
**Dev container name**: 

- universal

**Description**:

This PR addresses the GHSA-c2qf-rxjj-qqgw vulnerability. The vulnerability is related to the `smever` package. This package is a dependency for `npm` shipped with Node v18.17.0.

*Changelog*:

- Bumped npm to the `9.8.1` version;
- Added test to verify npm version;

*Other changes*:

- Locked the `dotnet` version to resolve the issue with Oryx;
- The `patch-python` feature updated:
  - Removed the `setuptools` patch for Python 3.10 since not required anymore;
  - Locked versions for packages;

**Checklist**:

- [x] Checked that applied changes work as expected

### CODE REVIEW

File: src/universal/.devcontainer/local-features/patch-python/install.sh [hunk 1]
```diff
@@ -41,8 +41,7 @@ update_package() {
 # They are installed by the base image (python) which does not have the patch.
 
 # https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2022-40897
-update_package /usr/local/python/3.9.*/bin/python setuptools
-update_package /usr/local/python/3.10.*/bin/python setuptools
+update_package /usr/local/python/3.9.*/bin/python setuptools==65.5.1
```
Comment (samruddhikhandale) [inline src/universal/.devcontainer/local-features/patch-python/install.sh:44]: What's the current version of `setuptools ` installed in the prod image? 

From https://github.com/devcontainers/images/blob/main/src/universal/history/dev.md , `/usr/local/python/3.10.*/bin/python*` has `68.0.0`. We should pin to the version which is currently available in prod (this would avoid folks running into issues with degrading tool versions)

File: src/universal/.devcontainer/local-features/setup-user/install.sh [hunk 2]
```diff
@@ -97,4 +97,10 @@ find "${OPT_DIR}" -type d | xargs -n 1 chmod g+s
 
 echo "Defaults secure_path=\"${DOTNET_PATH}:${NODE_PATH}/bin:${PHP_PATH}/bin:${PYTHON_PATH}/bin:${JAVA_PATH}/bin:${RUBY_PATH}/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/local/bin:/usr/local/share:/home/${USERNAME}/.local/bin:${PATH}\"" >> /etc/sudoers.d/$USERNAME
 
+# Temporary: Due to GHSA-c2qf-rxjj-qqgw
+source "${NVM_DIR}/nvm.sh"
+nvm use 18
+npm -g install -g npm@9.8.1
+nvm use stable
```
Comment (samruddhikhandale) [inline src/universal/.devcontainer/local-features/setup-user/install.sh:104]: Would `nvm use 18` and `nvm use stable` have unintentional impacts on the dev container?

File: src/universal/test-project/test.sh [hunk 3]
```diff
@@ -53,7 +53,7 @@ check "seaborn" python -c "import seaborn; print(seaborn.__version__)"
 check "scikit-learn" python -c "import sklearn; print(sklearn.__version__)"
 check "torch" python -c "import torch; print(torch.__version__)"
 check "requests" python -c "import requests; print(requests.__version__)"
-check "jupyterlab-git" bash -c "python3 -m pip list | grep jupyterlab-git"
+check "jupyterlab-git" python -c "import jupyterlab_git; print(jupyterlab_git.__version__)"
```
Comment (samruddhikhandale) [inline src/universal/test-project/test.sh:56]: Interesting, did the maintainers rename the `jupyterlab-git` package?

Comment (alexander-smolyakov) [reply src/universal/test-project/test.sh:56]: It seems so. It looks like the package was renamed in this PR:
- jupyterlab/jupyterlab-git/pull/1248

File: src/universal/.devcontainer/local-features/setup-user/install.sh (hunk 2 above)
Comment (alexander-smolyakov) [reply src/universal/.devcontainer/local-features/setup-user/install.sh]: > `nvm use 18`

We have to switch the Node version to install the new `npm` version globally correctly. After that, the version switched to the default to preserve the original lineup. I don't see any side effects for the devcontainer. Please let me know if I'm missing something. 

> `nvm use stable`

The `default` alias is set up during [Node feature](https://github.com/devcontainers/features/tree/main/src/node) installation and [references the version specified](https://github.com/devcontainers/features/blob/9e75db9a1dbcc245f94707c0917470143c17595b/src/node/install.sh#L10) in the `version` property.

_Related code_:

- The `default` alias is set up: [node/install.sh](https://github.com/devcontainers/features/blob/9e75db9a1dbcc245f94707c0917470143c17595b/src/node/install.sh#L123C1-L125C3);
- The `Node` feature definition for `universal` devcontainer: [devcontainer.json](https://github.com/devcontainers/images/blob/d394ecd84617214fb5aa086a2b5a27704e5c868d/src/universal/.devcontainer/devcontainer.json#L20C1-L23C11);

To summarize everything, the final flow looks like this:
- switch from Node 20 to Node 18
- install new npm version for Node 18 
- switch back from Node 18 to Node 20

File: src/universal/.devcontainer/local-features/patch-python/install.sh (hunk 1 above)
Comment (alexander-smolyakov) [reply src/universal/.devcontainer/local-features/patch-python/install.sh:44]: Good point! Locked version in: https://github.com/devcontainers/images/pull/700/commits/0d0c6dde1e61129386fa70d924732d6fb83b5a2c

Comment (alexander-smolyakov) [reply src/universal/.devcontainer/local-features/patch-python/install.sh:44]: Currently investigating permission denied error:

```console
2023-08-11T15:00:10.9105821Z [2023-08-11T15:00:10.910Z] #44 2.762 Uninstalling setuptools-65.5.1:
2023-08-11T15:00:10.9106218Z #44 2.774 ERROR: Exception:
2023-08-11T15:00:10.9106542Z #44 2.774 Traceback (most recent call last):
2023-08-11T15:00:10.9106841Z #44 2.774   File "/opt/python/3.10.8/lib/python3.10/shutil.py", line 815, in move
2023-08-11T15:00:10.9107550Z #44 2.774     os.rename(src, real_dst)
2023-08-11T15:00:10.9108422Z #44 2.774 PermissionError: [Errno 13] Permission denied: '/opt/python/3.10.8/lib/python3.10/site-packages/_distutils_hack/' -> '/tmp/pip-uninstall-q5woibce'
2023-08-11T15:00:10.9108759Z #44 2.774 
2023-08-11T15:00:10.9109180Z #44 2.774 During handling of the above exception, another exception occurred:
2023-08-11T15:00:10.9109457Z #44 2.774 
2023-08-11T15:00:10.9109782Z #44 2.774 Traceback (most recent call last):
2023-08-11T15:00:10.9110269Z #44 2.774   File "/opt/python/3.10.8/lib/python3.10/site-packages/pip/_internal/cli/base_command.py", line 180, in exc_logging_wrapper
2023-08-11T15:00:10.9110591Z #44 2.774     status = run_func(*args)
2023-08-11T15:00:10.9111016Z #44 2.774   File "/opt/python/3.10.8/lib/python3.10/site-packages/pip/_internal/commands/uninstall.py", line 105, in run
2023-08-11T15:00:10.9111336Z #44 2.774     uninstall_pathset = req.uninstall(
2023-08-11T15:00:10.9111765Z #44 2.774   File "/opt/python/3.10.8/lib/python3.10/site-packages/pip/_internal/req/req_install.py", line 687, in uninstall
2023-08-11T15:00:10.9112110Z #44 2.774     uninstalled_pathset.remove(auto_confirm, verbose)
2023-08-11T15:00:10.9112559Z #44 2.774   File "/opt/python/3.10.8/lib/python3.10/site-packages/pip/_internal/req/req_uninstall.py", line 381, in remove
2023-08-11T15:00:10.9112881Z #44 2.774     moved.stash(path)
2023-08-11T15:00:10.9113283Z #44 2.774   File "/opt/python/3.10.8/lib/python3.10/site-packages/pip/_internal/req/req_uninstall.py", line 272, in stash
2023-08-11T15:00:10.9113575Z #44 2.774     renames(path, new_path)
2023-08-11T15:00:10.9113935Z #44 2.774   File "/opt/python/3.10.8/lib/python3.10/site-packages/pip/_internal/utils/misc.py", line 315, in renames
2023-08-11T15:00:10.9114163Z #44 2.774     shutil.move(old, new)
2023-08-11T15:00:10.9114375Z #44 2.774   File "/opt/python/3.10.8/lib/python3.10/shutil.py", line 833, in move
2023-08-11T15:00:10.9114631Z #44 2.774     rmtree(src)
2023-08-11T15:00:10.9114837Z #44 2.774   File "/opt/python/3.10.8/lib/python3.10/shutil.py", line 724, in rmtree
2023-08-11T15:00:10.9115049Z #44 2.774     _rmtree_safe_fd(fd, path, onerror)
2023-08-11T15:00:10.9115282Z #44 2.774   File "/opt/python/3.10.8/lib/python3.10/shutil.py", line 657, in _rmtree_safe_fd
2023-08-11T15:00:10.9115565Z #44 2.774     _rmtree_safe_fd(dirfd, fullname, onerror)
2023-08-11T15:00:10.9115796Z #44 2.774   File "/opt/python/3.10.8/lib/python3.10/shutil.py", line 680, in _rmtree_safe_fd
2023-08-11T15:00:10.9116031Z #44 2.774     onerror(os.unlink, fullname, sys.exc_info())
2023-08-11T15:00:10.9116267Z #44 2.774   File "/opt/python/3.10.8/lib/python3.10/shutil.py", line 678, in _rmtree_safe_fd
2023-08-11T15:00:10.9116486Z #44 2.774     os.unlink(entry.name, dir_fd=topfd)
2023-08-11T15:00:10.9116798Z #44 2.774 PermissionError: [Errno 13] Permission denied: 'override.cpython-310.pyc'
2023-08-11T15:00:10.9117078Z #44 2.809 ERROR: Feature "Patch Python Packages" (Unknown) failed to install!
```

Comment (alexander-smolyakov) [reply src/universal/.devcontainer/local-features/patch-python/install.sh:44]: **Investigation summary**: 
The Issue related to the Python folder having the `root root` permissions, which caused the `permissions denied` error.

After recent changes in `devcontainer cli`, the Oryx feature now installs before the Python feature (However earlier Oryx feature was installed after the Python feature according to [installation queue](https://github.com/devcontainers/images/blob/6aee5de3e32e539a3d7c0ae158845650231aff47/src/universal/.devcontainer/devcontainer.json#L78C1-L102)). Because of this, all Pythons distribution now [installs via Oryx](https://github.com/devcontainers/features/blob/f90cb26c7f15659f3e2be8061295997df2bb76cc/src/python/install.sh#L351-L352), and starting from some point, we are facing a `permissions denied` error.

The current fix is to explicitly set the `useOryxIfAvailable` [option](https://github.com/devcontainers/features/blob/f90cb26c7f15659f3e2be8061295997df2bb76cc/src/python/install.sh#L20C1-L20C22) for the Python feature to `false`, to preserve the original installation process. 

**Next mitigation steps**:

- Check the Python feature to investigate why installing Python distribution via Oryx could cause a `permissions denied` error;
- Discuss the Python installation process for universal devcontainer. Is using Oryx to install Python fine, or is it better to preserve the original behavior?

Comment (alexander-smolyakov) [reply src/universal/.devcontainer/local-features/patch-python/install.sh:44]: **Attaching screenshot for context**:
_Left_: Permissions for Python 3.10.8
_Right_: Permissions for Python 3.10.12

Both are installed via Oryx. It looks like there is an issue either with Oryx or Python.

image

File: src/universal/.devcontainer/devcontainer.json [hunk 4]
```diff
@@ -10,7 +10,7 @@
             "userGid": "1000"
         },
         "ghcr.io/devcontainers/features/dotnet:1": {
-            "version": "7",
+            "version": "7.0.306",
```
Comment (alexander-smolyakov) [inline src/universal/.devcontainer/devcontainer.json:13]: Related issue (for context): 
- microsoft/Oryx/issues/2094

File: src/universal/.devcontainer/local-features/patch-python/install.sh (hunk 1 above)
Comment (samruddhikhandale) [reply src/universal/.devcontainer/local-features/patch-python/install.sh:44]: > Is using Oryx to install Python fine, or is it better to preserve the original behavior?

I think it's better to preserve the old behavior and avoid using oryx to install Python 👍

**Review by samruddhikhandale — APPROVED** [review]
